front/lib/api/assistant/generation.ts

```diff
@@ -175,16 +180,51 @@ export async function renderConversationForModel({
   let tokensUsed = promptCountRes.value + 64;
 
   // Go backward and accumulate as much as we can within allowedTokenCount.
-  const selected = [];
+  const selected: ModelMessageType[] = [];
   for (let i = messages.length - 1; i >= 0; i--) {
+    if (tokensUsed >= allowedTokenCount) {
+      break;
+    }
     const r = messagesCountRes[i];
     if (r.isErr()) {
       return new Err(r.error);
     }
     const c = r.value;
     if (tokensUsed + c <= allowedTokenCount) {
       tokensUsed += c;
-      selected.unshift(messages[i]);
+      selected.unshift({
+        role: messages[i].role,
+        content: messages[i].content,
+        name: messages[i].name,
+      });
+    } else if (messages[i].role === "content_fragment") {
+      const c = allowedTokenCount - tokensUsed;
+      const contentFragmentMessage = messages[i].message;
+      if (contentFragmentMessage) {
+        if (isContentFragmentType(contentFragmentMessage)) {
+          switch (contentFragmentMessage.contentType) {
+            case "file_attachment":
+              // Truncate file attachment content to the first characters.
+              messages[i].content = messages[i].content.substring(0, c);
+              break;
+            case "slack_thread_content":
+              // Truncate slack thread content to the last characters.
+              messages[i].content = messages[i].content.substring(
+                messages[i].content.length - c
+              );
+              break;
+            default:
+              assertNever(contentFragmentMessage.contentType);
+          }
+        }
+      }
+
+      tokensUsed += c;
```

Comment:
Let's move this inside the double if and make the double if one if for clarity :pray: